### DISCUSSION TITLE
help with calysto_hy updating

### DISCUSSION THREAD

**Discussion by actsasgeek** [hylang/hy#2362] [Q&A]
I decided to try and tackle an update to calysto_hy, because using Jupyter notebooks is pretty much _required_ for my use cases and the repo hasn't been updated for several years. 

I don't really know what I'm doing but I'm willing to be educated. After futzing around a bit, I noticed that things were pretty big changes between hy 0.24 and 0.25. Conceptually, I think hy 0.25 greatly simplifies the task but again there's not much (any?) documentation for Hy's architecture, MetaKernel's docs are out of date.

I replaced this block:

https://github.com/Calysto/calysto_hy/blob/master/calysto_hy/kernel.py#L136-L146

with:

```
        with filtered_hy_exceptions():
            self.result = self.repl.runsource(code, f"In [{self.execution_count}]", "single")
```

and it actually "works"...if you're an expert programmer who never makes mistakes. However, I can't figure out how to report errors. I thought I could something like this:

https://github.com/hylang/hy/blob/master/hy/cmdline.py#L58-L65

but the return value is never anything by False for me.

At this point, I figured I might need some help understanding what is actually happening. 

1. What is supposed to be returned from `runsource`? 
2. How can I access the error information if an error occurred? I notice it prints it out so...
3. quite possibly, I can't use REPL "as is" but I need to do what it is doing in Calysto Hy.

I noticed that TryHy subclassed REPL but I couldn't quite suss out what it was doing.

**Comment by Kodiologist** [comment]
> there's not much (any?) documentation for Hy's architecture

Yeah, Hy internals aren't considered part of the public interface, so we may break anything that depends on them without notice.

> I noticed that things were pretty big changes between hy 0.24 and 0.25

That's surprising seeing as I've made all the code changes since the release of 0.24.0, and I wouldn't consider any of them particularly big. But, I don't know anything about Jupyter or what it depends on.

**Reply by actsasgeek** [reply]
Thanks
